sdk/eventhub/Azure.Messaging.EventHubs.Processor/tests/Processor/EventProcessorClientTests.cs

```diff
@@ -700,113 +674,1998 @@ public async Task IsRunningReturnsTrueWhileStopProcessingAsyncIsNotCalled()
+                Assert.That(eventArgs.Reason, Is.EqualTo(ProcessingStoppedReason.Shutdown), $"The partition '{ eventArgs.PartitionId }' should have stopped processing with the correct reason.");
+            }
+        }
+
+        /// <summary>
+        ///   Verifies functionality of the <see cref="EventProcessorClient.PartitionClosingAsync" />
+        ///   event.
+        /// </summary>
+        ///
+        [Test]
+        [Ignore("Flaky test. (Tracked by: #10015)")]
+        public async Task PartitionClosingAsyncIsCalledWithOwnershipLostReasonWhenStoppingTheFailedProcessor()
+        {
+            var mockConsumer = new Mock<EventHubConsumerClient>("consumerGroup", Mock.Of<EventHubConnection>(), default);
+            var mockProcessor = new InjectableEventSourceProcessorMock(new MockCheckPointStorage(), "consumerGroup", "namespace", "eventHub", Mock.Of<Func<EventHubConnection>>(), default, mockConsumer.Object);
+
+            var partitionIds = new[] { "0", "1" };
+            var faultedPartitionId = partitionIds.Last();
+
+            mockProcessor.CatchNUnitRunnerException = (partitionId, position, token) => partitionId != faultedPartitionId;
+
+            mockConsumer
+                .Setup(consumer => consumer.GetPartitionIdsAsync(It.IsAny<CancellationToken>()))
+                .Returns(Task.FromResult(partitionIds));
+
+            var completionSource = new TaskCompletionSource<bool>();
+            var partitionsBeingProcessed = 0;
+
+            // Use the ReadEventsFromPartitionAsync method to notify when all partitions have started being processed. If the partition is faulted,
+            // throw.
+
+            mockConsumer
+                .Setup(consumer => consumer.ReadEventsFromPartitionAsync(
+                    It.IsAny<string>(),
+                    It.IsAny<EventPosition>(),
+                    It.IsAny<ReadEventOptions>(),
+                    It.IsAny<CancellationToken>()))
+                .Callback(() =>
+                {
+                    if (Interlocked.Increment(ref partitionsBeingProcessed) == partitionIds.Length)
+                    {
+                        completionSource.SetResult(true);
+                    }
+                })
+                .Returns<string, EventPosition, ReadEventOptions, CancellationToken>((partition, position, options, token) =>
+                {
+                    if (partition == faultedPartitionId)
+                    {
+                        throw new Exception();
+                    }
+
+                    return MockEndlessPartitionEventEnumerable(options.MaximumWaitTime, token);
+                });
+
+            // Use the close handler to keep track of the partitions that have been closed.
+
+            var closingEventArgs = new ConcurrentDictionary<string, PartitionClosingEventArgs>();
+
+            mockProcessor.PartitionClosingAsync += eventArgs =>
+            {
+                closingEventArgs[eventArgs.PartitionId] = eventArgs;
```

Comment:
maybe try another completion source here to avoid the delay loop after?

Reply:
I've been using the delay loop in my tests so we don't end up hanging indefinitely in case of unexpected behavior. You can see I have multiple tests in which I'm using the completion source AND the delay loop. This flaky test, in particular, hangs sometimes if I don't use the loop, so I think leaving it like this is fine.

Reply:
If we're seeing consistent results across environments, I'm good with leaving it.   I've been slowly trying to root out the timing-based things as I babysit tests as they tend to be cranky when the nightly runs are under load and there's a lot of delays.